mr/src/main/java/org/apache/iceberg/mr/mapreduce/IcebergInputFormat.java

```diff
@@ -348,49 +342,38 @@ public void close() throws IOException {
       currentIterator.close();
     }
 
-    private static Map<String, Integer> buildNameToPos(Schema expectedSchema) {
-      Map<String, Integer> nameToPos = Maps.newHashMap();
-      for (int pos = 0; pos < expectedSchema.asStruct().fields().size(); pos++) {
-        Types.NestedField field = expectedSchema.asStruct().fields().get(pos);
-        nameToPos.put(field.name(), pos);
-      }
-      return nameToPos;
-    }
-
-    private CloseableIterator<T> open(FileScanTask currentTask) {
-      DataFile file = currentTask.file();
-      // schema of rows returned by readers
-      PartitionSpec spec = currentTask.spec();
-      Set<Integer> idColumns =  Sets.intersection(spec.identitySourceIds(), TypeUtil.getProjectedIds(expectedSchema));
-      boolean hasJoinedPartitionColumns = !idColumns.isEmpty();
+    private CloseableIterator<T> open(FileScanTask task) {
+      PartitionSpec spec = task.spec();
+      Set<Integer> idColumns = spec.identitySourceIds();
+      Schema partitionSchema = TypeUtil.select(expectedSchema, idColumns);
+      boolean projectsIdentityPartitionColumns = !partitionSchema.columns().isEmpty();
 
       CloseableIterable<T> iterable;
-      if (hasJoinedPartitionColumns) {
-        Schema readDataSchema = TypeUtil.selectNot(expectedSchema, idColumns);
-        Schema identityPartitionSchema = TypeUtil.select(expectedSchema, idColumns);
-        iterable = CloseableIterable.transform(open(currentTask, readDataSchema),
-            row -> withIdentityPartitionColumns(row, identityPartitionSchema, spec, file.partition()));
+      if (projectsIdentityPartitionColumns) {
+        //TODO: seems like we have to specify a converter to convert the partition values
```

Comment:
I was hesitant to use the converter from TableScanIterable as it seems it might only supports Avro.

Reply:
Yes, we should use the one from `TableScanIterable`. That conversion is from internal representations (e.g., long for timestamps) to Iceberg generics (`LocalDateTime` or `OffsetDateTime` for timestamps) so it is the right one to use here. When we add support for Pig or Hive objects, we would need to convert differently for those.

Reply:
Any thoughts on which is the right place to put the converter? Or maybe I can just make the one in TableScanIterable public and use that?

Reply:
Now that the generics for Avro are in core, we can put them in the org.apache.iceberg.data package in iceberg-core. How about a new `Converters` class?

Reply:
Fixed